File: python/robot-simulator/robot_simulator.py
NORTH = 0
EAST = 1
SOUTH = 2
WEST = 3


class Robot(object):
    def __init__(self, bearing=NORTH, x=0, y=0):
        self.coordinates = (x, y)
        self.bearing = bearing
    
    def turn_right(self):
        self.bearing = (self.bearing + 1) % 4

    def turn_left(self):
        self.bearing = (self.bearing -1 + 4) % 4
    
    def advance(self):
        if self.bearing == NORTH:
            self.coordinates = (self.coordinates[0], self.coordinates[1] + 1)
        if self.bearing == EAST:
            self.coordinates = (self.coordinates[0] + 1, self.coordinates[1])
        if self.bearing == SOUTH:
            self.coordinates = (self.coordinates[0], self.coordinates[1] - 1)
        if self.bearing == WEST:
            self.coordinates = (self.coordinates[0] - 1, self.coordinates[1])

    def simulate(self, commands):
        for command in commands:
            if command == 'R':
                self.turn_right()
            elif command == 'L':
                self.turn_left()
            else:
                self.advance()
            

File: javascript/robot-simulator/robot-simulator.js
const DIRECTIONS = ['north', 'east', 'south','west'];
const INSTRUCTIONS = {
  L: 'turnLeft',
  R: 'turnRight',
  A: 'advance',
}

export class InvalidInputError extends Error {
  constructor(message) {
    super(message);
  }
}

export class Robot {
  static instructions(instructions) {
    const listOfInstructions = []
    for (const instruction of instructions) {
      listOfInstructions.push(INSTRUCTIONS[instruction])
    }
    return listOfInstructions;
  }

  orient(bearing) {
    if (!DIRECTIONS.includes(bearing)) {
      throw new InvalidInputError('Invalid bearing');
    }
    this.bearing = bearing;
  }

  turnRight() {
    this.bearing = DIRECTIONS[(DIRECTIONS.indexOf(this.bearing) + 1) % 4];
  }

  turnLeft() {
    this.bearing = DIRECTIONS[(DIRECTIONS.indexOf(this.bearing) - 1 + 4) % 4];
  }

  at(x, y) {
    this.x = x;
    this.y = y;
    this.coordinates = [x, y];
  }

  advance() {
    if (this.bearing === 'north') this.coordinates = [this.coordinates[0], this.coordinates[1] + 1];
    else if (this.bearing === 'east') this.coordinates = [this.coordinates[0] + 1, this.coordinates[1]];
    else if (this.bearing === 'south') this.coordinates = [this.coordinates[0], this.coordinates[1] - 1];
    else this.coordinates = [this.coordinates[0] - 1, this.coordinates[1]];
  }

  place(info) {
    this.coordinates = [info.x, info.y];
    this.bearing = info.direction;
  }

  evaluate(instructions) {
    const listOfInstructions = Robot.instructions(instructions);
    for (const instruction of listOfInstructions) {
      this[instruction]();
    }
  }
}
